feat: Initial relay implementation

bellini666 opened the pull request:
This is my relay implementation from [strawberry-django-plus](https://github.com/blb-ventures/strawberry-django-plus), which I mentioned in [this issue](https://github.com/strawberry-graphql/strawberry/issues/1573)

There are some things that might need some discussion:

1. I used `GlobalID` as the name of the scalar to avoid conflicts with the existing `ID`. Although it doesn't cause issues for the majority of people, [there are some cases](https://github.com/blb-ventures/strawberry-django-plus/issues/166) where the client expects the scalar to be named `ID`. It can be worked around by overriding the `ID` scalar to use the `GlobalID`, but maybe there's a better way? Note that the `GlobalID` object is a dataclass with some custom attributes and methods.
2. Find a way to allow the object to define an `id`, which is the non global version one, while still keeping our `Node.id` resolver working. You will notice in my tests that I defined the `Fruit` id as `_id` for that reason. Maybe use a metaclass approach to solve this?

TODO (already done):
- [x] Add tests for the `input_mutation`
- [x] Write a documentation for `input_mutation`

Fix #1573

botberry commented:
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Thiago Bellini Ribeiro for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```
<!-- action-check: release-file -->

kaos commented on pyproject.toml:
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "strawberry-graphql"
-packages = [ { include = "strawberry" } ]
+packages = [{ include = "strawberry" }]
```
I discourage formatting changes mixed with other changes. Particularly for a PR already of this magnitude. It makes it so much more difficult to see what's actually changed.

bellini666 replied:
Oh, totally! I'm also the one who comments that on PRs I review. My autoformatter formatted this when I modified this file and I didn't even notice =P

kaos commented on strawberry/relay/types.py:
```diff
@@ -0,0 +1,706 @@
+_R = TypeVar("_R")
+connection_typename = "arrayconnection"
+
+NodeType = TypeVar("NodeType", bound="Node")
```
This looks like a class name to me, and reading the docs example for custom connections had me really confused (wrgt how it was implemented, not how to use it).

I'm not sure how to improve this really, though. Perhaps `TNode` does a better job at hinting that this is a generic type var?

however, looking at some more places this is used, I think `NodeType` works really well, so perhaps the doc mentioned is the only confusing part.. I'll ponder on this a bit more.

bellini666 replied:
I'm open to change it to `TNode` or anything else that makes more sense. I have to admit that I'm not good at naming things (`strawberry-django-plus` is a good example of that hahaha)

kaos replied:
Naming stuff is among the top 3 most difficult programming jobs! :P 

I think I landed in being in favour of keeping it as `NodeType`, left the comment regardless to be transparent about my review process.

kaos commented on tests/relay/schema.py:
```diff
@@ -0,0 +1,185 @@
+    @classmethod
+    def from_nodes(
+        cls,
+        nodes: Iterable[Fruit],
```
This should be `Iterable[NodeType]` given that the type is not specialized here, but still a `Generic[NodeType]`, which is wrong as we want it to be specialized now.

I think the correct use here is:
```python
@strawberry.type
class CustomPaginationConnection(
    strawberry.relay.Connection[Fruit]
):
    ...


@strawberry.type
class Query:
    ...
    fruits_custom_pagination: CustomPaginationConnection

```

This would explain my confusion I had with this, and also it would then make sense to perhaps call it `CustomFruitPaginationConnection` ;)

bellini666 replied:
Yeah, I agree. I actually tried to do that when I originally wrote this test on `strawberry-django-plus`, but strawberry was not able to generate the type that way.

Maybe now that this code is being merged here I can try to modify the Generic handling to support that code.

bellini666 replied:
I investigated and discovered the issue here. It was not specific to the relay implementation, but with `Generic` type trying to create a specialized type in that way.

Basically, when creating the fields, the `type_var_map` was not being set for them. I fixed the issue and added tests for it in `test_generics`: https://github.com/strawberry-graphql/strawberry/pull/2511/commits/6bc2f4f0b4f311125144c64f1aa07cff8ec9fa06

kaos reviewed (approved):
This is a massive PR, but I hope we can get through this as it looks really well worked through. 🚀 

And it will be easier to circle back and fix any remaining rough edges after we've got some experience using it.

kaos commented on docs/guides/relay.md:
```diff
@@ -263,9 +261,18 @@ class CustomPaginationConnection(
 
+<Note>
+
+In the example above we specialized the `FruitCustomPaginationConnection` by
```
```suggestion
In the example above we specialized the `CustomPaginationConnection` by
```

kaos commented on docs/guides/relay.md:
```diff
@@ -263,9 +261,18 @@ class CustomPaginationConnection(
+In the example above we specialized the `FruitCustomPaginationConnection` by
+inheriting it from `relay.Connection[Fruit]`. We could still keep it generic by
+inheriting it from `relay.Connection[relay.NodeType]` and then specialize it
+in when defining the field.
```
```suggestion
when defining the field.
```

kaos reviewed (approved):
Nice!

bellini666 commented:
Just added a doc section for the input mutation and some tests for it. I think the PR is basically done, not considering any suggestions/improvements that might arrive from code reviews, of course :)

bellini666 commented:
Note: I separated the specialized generic support to another PR (https://github.com/strawberry-graphql/strawberry/pull/2517). As soon as it gets merged I'll rebase this PR to make the changes here be only about relay itself

bellini666 commented:
> Note: I separated the specialized generic support to another PR (#2517). As soon as it gets merged I'll rebase this PR to make the changes here be only about relay itself

It was merged! Just merged main in here, the diff should now only contain relay related stuff

patrick91 commented:
/pre-release

botberry commented:
# Pre-release
:wave:

Releasing commit [5d582dbf43e2084603f3b4ec3c401ba1b5991c7f] to PyPi as pre-release! :package:

botberry commented:
# Pre-release
:wave:

Pre-release **0.157.0.dev.1675982393** [5d582dbf43e2084603f3b4ec3c401ba1b5991c7f] has been released on PyPi! :rocket:
You can try it by doing:

```shell
poetry add strawberry-graphql==0.157.0.dev.1675982393
```

patrick91 commented on strawberry/types/type_resolver.py:
```diff
@@ -50,6 +51,15 @@ class if one is not set by either using an explicit strawberry.field(name=...) o
     # Deferred import to avoid import cycles
     from strawberry.field import StrawberryField
 
+    # Relay uses @interface/@type while creating its types, which calls this function
+    try:
+        from strawberry.relay import Connection, ConnectionField, Node, NodeField
+    except ImportError:
+        Connection = None  # type: ignore[misc,assignment]
+        ConnectionField = None  # type: ignore[misc,assignment]
+        Node = None  # type: ignore[misc,assignment]
+        NodeField = None  # type: ignore[misc,assignment]
```
why would we get an import error here?

bellini666 replied:
Inside `relay/types.py` the types defined there use `@strawberry.type`, which imports this file to resolve their fields and causes a circular import error.

Probably my comment above is not enough to explain the issue here =P. But when creating the types there, `Node`, `Edge` and `Connection`, this will raise an `ImportError`. After that the import should work fine

patrick91 replied:
this looks like something we need to refactor, either in this PR or another one :D

bellini666 replied:
I refactored it to make the code better to understand

patrick91 commented on strawberry/relay/types.py:
```diff
@@ -0,0 +1,706 @@
+
+_T = TypeVar("_T")
+_R = TypeVar("_R")
+connection_typename = "arrayconnection"
```
could we avoid to use this somehow? I think we can always use the type name from strawberry, no? :)

bellini666 replied:
This actually is the name of the cursor (maybe it is poorly named? =P)

It is used in here: https://github.com/strawberry-graphql/strawberry/pull/2511/files#diff-87ff2a77ed8a7dc2eb7fc8a320e55d457b77160ab7932c185b8becbf8662767bR526 . I used the same name that [graphql-relay-py](https://github.com/graphql-python/graphql-relay-py/blob/97ba15e573eabcefd13e249d05fc5557f0296804/src/graphql_relay/connection/array_connection.py#L158) uses.

I probably wanted to call this `cursor_typename` but wrote `connection_typename` by mistake. Do you have any suggestions for a better name?

patrick91 replied:
oh I mean `arrayconnection`, wouldn't be best to use the name of the types we are going to return? we do have access to those, no?

bellini666 replied:
They are actually not used in type names, but in the cursor. For example:

```json
{
  "edges" [
    {
      "cursor": "arrayconnection:0",  # this
      "node": {
        "__typename": "Fruit",
        "name": "Banana"
      },
    },
    {
      "cursor": "arrayconnection:1",  # this
      "node": {
        "__typename": "Fruit",
        "name": "Apple"
      },
    },
  ]
}
```

It is used in those places, but it is exposed as `base64` (I put them raw in the example above to show where they are used).

The cursor has the pagination information, which is not necessarily related to the type itself. For the base implementation, which uses limit/offset, it will be the row number of the result. So, if your connection has a total number objects of `1000`, you would have cursors from `arrayconnection:0` to `arrayconnection:999`. Then when asking for `first: 10` and `after: arrayconnection:50` (but in base64) for example your results would have edges from `arrayconnection:51` to `arrayconnection:61`

So the first cursor in a connection of 2 different types will both be `arrayconnection:0` because they are both the rownumber 0 in that connection.

Regarding the name, it is also used in this example (https://relay.dev/docs/guides/graphql-server-specification/). If you decode the base64 inside the cursor, you can see that it also uses `arrayconnection` for it

patrick91 commented on strawberry/relay/mutations.py:
```diff
@@ -0,0 +1,232 @@
+    ...
+
+
+def input_mutation(
+    resolver=None,
+    *,
+    name: Optional[str] = None,
+    is_subscription: bool = False,
+    description: Optional[str] = None,
+    permission_classes: Optional[List[Type[BasePermission]]] = None,
+    deprecation_reason: Optional[str] = None,
+    default: Any = dataclasses.MISSING,
+    default_factory: Union[Callable[..., object], object] = dataclasses.MISSING,
+    metadata: Optional[Mapping[Any, Any]] = None,
+    directives: Optional[Sequence[object]] = (),
+    # This init parameter is used by pyright to determine whether this field
+    # is added in the constructor or not. It is not used to change
+    # any behavior at the moment.
+    init: Literal[True, False, None] = None,
+) -> Any:
```
let's move this into another PR and decide if we should keep it or not 😊

bellini666 replied:
Sure! Will do that

bellini666 replied:
Done! Will open another PR soon

Should it go to `strawberry/mutation.py` as an alternative mutation approach?

patrick91 commented on strawberry/relay/fields.py:
```diff
@@ -0,0 +1,553 @@
+class NodeField(RelayField):
+    """Relay Node field.
+
+    Do not instantiate this directly. Instead, use `@relay.node`
```
is there any example of this?

bellini666 replied:
You got me! hahaha

This is a very dated comment inherited from my code in strawberry-django-plus. Originally this field had the possibility to receive a resolver, but soon after I did that I noticed that it doesn't make sense to have one, because it is just a matter of creating a `@strawberry.field` with a single argument of `id: GlobalID` in it and removed the functionality.

I'll remove this part of the doc

patrick91 commented on strawberry/relay/fields.py:
```diff
@@ -0,0 +1,553 @@
+@overload
+def connection(
+    *,
+    resolver: Callable[[], _T],
```
should this be resolver type?

bellini666 replied:
Oh, didn't even notice that the typing is bad in here. WIll fix it

patrick91 commented on strawberry/relay/fields.py:
```diff
@@ -0,0 +1,553 @@
+    ...
+
+
+@overload
+def connection(
+    resolver: Union[StrawberryResolver, Callable, staticmethod, classmethod],
+    *,
+    name: Optional[str] = None,
+    is_subscription: bool = False,
+    description: Optional[str] = None,
+    permission_classes: Optional[List[Type[BasePermission]]] = None,
+    deprecation_reason: Optional[str] = None,
+    default: Any = dataclasses.MISSING,
+    default_factory: Union[Callable[..., object], object] = dataclasses.MISSING,
+    metadata: Optional[Mapping[Any, Any]] = None,
+    directives: Optional[Sequence[object]] = (),
+    graphql_type: Optional[Any] = None,
+) -> ConnectionField:
```
this might be annoying, and I'm happy to do it for, but would be possible to add some pyright tests?

I think there's some small issues with typing when using strict mode:

<img width="769" alt="CleanShot 2023-02-09 at 23 23 35@2x" src="https://user-images.githubusercontent.com/667029/217961441-5331b26d-a841-4498-92fb-babdecd513cc.png">

bellini666 replied:
Sure, I can totally do that! :)

Does that only happen in strict mode? I use pyright in basic mode (with some extra checks enabled) and did not see that before. Will write some tests and fix any typing issues that strict mode shows :)

patrick91 replied:
yup!

patrick91 commented on docs/guides/relay.md:
```diff
@@ -0,0 +1,381 @@
+    def resolve_id(cls, root: Self, *, info: Optional[Info] = None):
+        return root.code
+
+    @classmethod
+    def resolve_nodes(
+        cls,
+        *,
+        info: Optional[Info] = None,
+        node_ids: Optional[Iterable[str]] = None,
+    ):
+        if node_ids is not None:
+            return [fruits[nid] for nid in node_ids]
+
+        return list(fruits.values())
+
+    @classmethod
+    def resolve_node(
+        cls,
+        node_id: str,
+        *,
+        info: Optional[Info] = None,
+        required: bool = False,
+    ):
```
are these methods always required?

bellini666 replied:
Yes, basically. I even defined then as `abstractmethods` in the `Node` interface.

The idea here is that any type implementing `Node` should define 3 methods:

1) `resolve_id`, which should return the real id of the object to be used to generate the global id. I would love to allow the type to use an `id` directly, but then there's no way to have it use the resolver (at least not without some refactoring to allow that)

2) `resolve_nodes`, which is expected to return an iterable of available nodes so that the connection can paginate over it

3) `resolve_node`, which should return the node given its id. During the migration I though about removing this one, since we can basically call `resolve_nodes` and pass `node_ids=[node_id]`. What do you think?

bellini666 replied:
Also note that integrations can define those automatically.

For example, in strawberry-django-plus, when a type is created and it implements `Node`, [I inject those resolvers into the type with async-safe resolvers](https://github.com/blb-ventures/strawberry-django-plus/blob/main/strawberry_django_plus/type.py#L293) that queries the django model, and also applying the query optimizer, etc. So anyone using the django integration there can just inherit from `Node` and not worry about those methods (unless he wants to override them for some reason)

patrick91 replied:
> resolve_id, which should return the real id of the object to be used to generate the global id. I would love to allow the type to use an id directly, but then there's no way to have it use the resolver (at least not without some refactoring to allow that)

I think we should find a way to make this work 😊

For `resolve_nodes` and `resolve_node`, I think it would be nice to have only one, but at the same time I wonder if `resolve_nodes` is needed all the time, for example if you do something like this:

```python
@strawberry.type
class Website:
    id: strawberry.ID
    domain: str

    @classmethod
    def from_model(cls, model: models.Website) -> Self:
        return cls(id=model.pk, domain=model.domain)

    @strawberry.relay.connection
    async def feedbacks(
        self, info: Info[Context, None]
    ) -> strawberry.relay.Connection[Feedback]:
        feedbacks = await find_feedbacks_for_site(self.domain)

        return strawberry.relay.Connection(
            edges=[
                strawberry.relay.Edge.from_node(Feedback.from_model(feedback))
                for feedback in feedbacks
            ],
            page_info=strawberry.relay.PageInfo(
                has_next_page=False,
                has_previous_page=False,
                start_cursor="",
                end_cursor="",
            ),
        )
```

I'm already passing the nodes to the connection, right? happy to chat about this on discord too :)

bellini666 replied:
> I think we should find a way to make this work :blush:

![download (3)](https://user-images.githubusercontent.com/134025/218112372-48e0b706-b849-4c62-b997-3b67cbbfa075.png)

> I'm already passing the nodes to the connection, right?

In that specific case I would argue that you could use `@strawberry.field` directly, because you are not using any of the "magic" behind `@strawberry.relay.connection`

The main use cases for the connection field are those:

1) Define the field as `feedbacks: Connection[Feedback]`. The field would be created with the `first`, `last`, `before` and `after` arguments, call `resolve_nodes` and use those arguments to properly paginate it.

2) Use `@strawberry.relay.connection` to return an `Iterable[Feedback]`, which is basically your way of giving the paginator your own nodes so that it doesn't have to call `resolve_nodes`. In the resolver you could even define some extra args for filtering/ordering the results, but forget about the pagination stuff (which is the `Connection`'s responsibility). Just like this example in the tests: https://github.com/strawberry-graphql/strawberry/pull/2511/files#diff-089578b7231144ec508d0023bb9562b896c8ef0e24ed10eabadfe7409028886dR161

A real example for this that I use: I have a django type that can be soft deleted. Since `resolve_nodes` would return the queryset with all objects, even those that were "deleted", i pre-filter the deleted ones in the resolver. For example:

```python
@strawberry.relay.connection
def fruits(self, include_removed: bool = False) -> Iterable[FruitType]:
    return Fruit.objects.filter(is_removed=not include_removed).order_by("name")
```

This would still generate this in the schema:

```graphql
type Query {
  fruits (
    first: Int
    last: Int
    before: String
    after: String
    includeRemoved: Bool! = false
  ): FruitTypeConnection!
}
```

So paginating over that connection would retrieve only results that are not removed (unless passing `includeRemoved: true`) and ordered by name

But having said all of that, I can totally not make them "mandatory", specially because the user can always use just the resolver approach, meaning `resolve_nodes` would not be used in this case. I can probably check when creating the schema that, when using the approach number 1 the `Node` implemented type has a `resolve_nodes` and give a warning to the user. What do you think?

> happy to chat about this on discord too :)

Sure! Always open to chat about anything on discord :)

bellini666 replied:
All resolved

patrick91 commented on tests/relay/schema.py:
```diff
@@ -0,0 +1,209 @@
+    color: str
+
+    @classmethod
+    def resolve_id(cls, root: Self, *, info: Optional[Info] = None):
```
is there a reason why info is optional?

bellini666 replied:
I thought that someone might want to call this somewhere they don't have access to an `info` object, and objects usually don't require it to resolve their IDs. But it is still there for objects that might need them

I can remove the optional if you prefer and make it required

kaos replied:
+1 on making it non-optional. It's easy enough to simply call another method from this one if you need to share the implementation with another callee that doesn't have `Info`.

In my use-case, Info is required, and marking it Optional only adds to my implementation to ensure it's not None to satisfy type checking.

patrick91 commented on tests/relay/schema.py:
```diff
@@ -0,0 +1,209 @@
+        cls,
+        *,
+        info: Optional[Info] = None,
+        node_ids: Optional[Iterable[str]] = None,
```
when would get get no `node_ids`?

bellini666 replied:
The main situation is: The `NodeField` have `id` or `ids` depending on whether it is a list or not. For example:

```python
@strawberry.type
class Query:
    node: Node
    nodes: List[Node]
```

This will generate:

```graphql
type Query {
  node(id: GlobalID!): Node!
  nodes(ids: [GlobalID!]!): [Node!]!
}
```

When querying `nodes` the list of global IDs will be passed in there

But it can be useful as a utility function for end users too.

bellini666 replied:
Oh, actually you question is when we would "not" get those.

When resolving a connection for example. The connection asks the `Node` for his nodes and paginate them. In this situation there isn't a list of nodes to be filtered, it just wants an iterable of "everything" (not actually everything, usually will be a lazy object that can be sliced) to slice.

patrick91 commented on docs/guides/relay.md:
```diff
@@ -0,0 +1,381 @@
+        )
+
+
+@strawberry.type
+class Query:
+    fruits: FruitCustomPaginationConnection
```
this is neat, but it will prevent to instantiate the type (query in this case, but can be another object type) without passing `fruits` :)

a workaround I found is this, but it's not nice:

```python
@strawberry.type
class MyObjectType:
    fruits: FruitCustomPaginationConnection = None
```

patrick91 replied:
I don't know if there's a solution for this other than maybe allowing:

```python
@strawberry.type
class MyObjectType:
    fruits: FruitCustomPaginationConnection = strawberry.relay.connection()
```

patrick91 commented on tests/relay/schema.gql:
```diff
@@ -0,0 +1,189 @@
+  edges: [FruitAsyncEdge!]!
+
+  """Total quantity of existing nodes"""
+  totalCount: Int
```
can we remove this field by default? I don't think it's in the spec 😊
Maybe we can have specialised connections for this?

patrick91 commented on docs/guides/relay.md:
```diff
@@ -0,0 +1,381 @@
+    ):
+        # Note that this is a showcase implementation and is far from
+        # being optimal performance wise
+        edges_mapping = {
+            to_base64("fruit_name", n.name): strawberry.relay.Edge(
+                node=n,
+                cursor=to_base64("fruit_name", n.name),
+            )
+            for n in sorted(nodes, key=lambda f: f.name)
+        }
```
this is the best place for allowing to customise edges?

bellini666 replied:
Currently, yes

bellini666 commented:
Just pushed a lot of fixes/refactors. The major highlights are:

* The `id` of the type implementing `Node` can now be `id: <something>`. For example an int id can be defined as `id: relay.NodeID[int]` and the implementation will know that this field is the id. This means we can also do `pk: relay.NodeID[int]` for example without having to override `resolve_id`. Also is explained in the documentation

* The nodefield can be defined as `node: Node`, `node: Node | None`, `node: list[Node]` or `node: list[Node | None]` and it will resolve them accordingly. More details in the documentation that I wrote for this

* The only "required" method to define in a type inheriting from `Node` now is `resolve_nodes`. It is not actually mandatory because if someone just uses it for connections with custom resolvers and also doesn't have any node fields, then this method will not be used. It is mandatory for anyone defining a generic `Node` (any of the 4 ways above) or a `Connection` field without a custom resolver

* Improved typing for the `connection` decorator/dataclass descriptor to fix pyright issues

* Removed `InputMutation`. I'll open another PR with it tomorrow. I'm thinking about making it generic inside `strawberry.mutation.py`, as an alternative approach to the default mutation. What do you think?

cc @patrick91

patrick91 commented:
/pre-release

bellini666 commented:
More changes:

* Add support for node resolvers return async generators
* Removal of `totalCount`
* Fixes for the limit/offset pagination logic
* New `node_converter` argument that can be passed to a `ConnectionField` to convert the node returned by the nodes resolver into its correct type when needed

kaos commented:
Are there a timeline/plan for this and/or is there a pre-release with the latest changes? If so I could try them out to help smoke out any surprise issues..

patrick91 commented:
@kaos we might want to merge this pr #2567 before ^^

I'll make a pre-release now, so you can test things 😊

patrick91 commented:
/pre-release

kaos commented:
> 

awesome, will try it out this weekend.

jthorniley commented on docs/guides/relay.md:
```diff
@@ -0,0 +1,394 @@
+
+
+# Assume we have a dict mapping the fruits code to the Fruit object itself
+fruits: Dict[int, Fruit]
```
I think the type annotation here should be `Dict[str, Fruit]` - because the node ID field in the example is str.

bellini666 replied:
`int` is actually fine because the globalid will do `str(id)` on it. Most ids are usually ints/uuids

bellini666 replied:
> So I think what I mean is either the dictionary should be str -> Fruit or the resolve_nodes class should have an int(nid) cast before using it as a key.

Indeed if the id was set to `str` it would work out of the box. But we can't just do `int(nid)` when passing the ids to `resolve_nodes` because we can't actually know the type of the ids when we resolve them.

e.g. You might have `FooType:1` and that `1` might be an integer or really a string containing the character `"1"`.

I don't know if there's a better way to workaround this, but I decided to go the easy way in this initial implementation and let the implementer cast the value. In this case, since you know that the id is an integer, you can do `int(nid)` inside the `resolve_nodes` function, like: `[fruits[int(nid)] if required else fruits.get(int(nid)) for nid in node_ids]`

But I'm open to suggestions

jthorniley replied:
I think the way you have it in the interface is fine. I don't think relay specifies any particular format for the ID so as you say you can't know.  Its fine to leave it up to the implementer.

I just mean in the example in the docs it would make it clearer to either have the `int(nid)` cast in the code, or make the dict keys `str` - as it is type checkers will error on the lookup. I've just noticed in the very similar code in the tests - `schema.py` line 122, the ids are cast to str in the lookup dictionary which makes sense.

bellini666 replied:
Oh, regarding the docs you are totally right, will fix it! Thanks for pointing it out

bellini666 replied:
Fixed!

jthorniley commented:
Hi, first off thanks for creating this it looks like a really nice addition! I've been looking into using it for our project based off the branch and it should be really useful.

I just wanted to comment on the `GlobalID`/`ID` issue.

I'm trying to migrate our project from graphene to strawberry and we have *many* Node types which I can't migrate all at once. I'm using the graphene/strawberry schema code linked from [this issue](https://github.com/strawberry-graphql/strawberry/issues/1042), which generally allows us to have both graphene and strawberry types in the schema.

So having the node interface change from the type being called `ID` to `GlobalID` is potentially a bit of an issue (assuming we get past any issues with compiling JS types), as our graphene types annotate the id field as `ID` not `GlobalID` (schema validation fails because the two kinds of "Node" think they are talking about different types, and generally I'd like to avoid schema changes as much as possible).

I've experimented with just setting the annotations to `ID` instead of `GlobalID` and doing some trivial manipulations to parse the types when needed. This seems to work for my case so I'm wondering if its a viable approach? The `GlobalID` dataclass can still be used as a utility to do the conversion between encoded string and type/id pair. 

My changes are [here](https://github.com/jthorniley/strawberry/compare/589d5ac3949a368387c73072399fe78e668353ab...jthorniley:strawberry:relay) - but it currently fails a lot of pytests fairly unsurprisingly, I"m just wondering if this seems like a good approach or are there other issues? (I haven't investigated how this affects connections at all yet).

bellini666 commented:
@patrick91 I did a major refactor of this code and added almost 100% coverage to it. If you can make a new pre-release from it to test (and also allow other people following this PR to do the same)

Note: After the refactoring and tests I tried to convert the `NodeField`/`ConnectionField` implementation to use `FieldExtension`. I made it work nicely (it is in [this commit](https://github.com/bellini666/strawberry/commit/7c662fd0fe78e860bfa18905da45ff6c15895777)), but there is a small issue that I wanted to discuss with you before moving forward:

In the commit I mentioned I ported `NodeField` to use `FieldExtension`. The implementation there is complete with the exception of [node_conveter](https://github.com/strawberry-graphql/strawberry/pull/2511/files#diff-ee83873f7394d1df47f446d9e9fe7a2e270a1c88ab3aa4478118d47e2f936af3R328). How can we pass `node_converter` to the node field without having to define a `node` function?

I was thinking in two possibilities:

1) Move the `node_converter` functionality to the `StrawberryField` itself as `result_converter` (or some other better name). It would allow the converter to be used for other fields as well, which could be actually good since it could resolve the issue with integrations where users usually have to rely on `cast` to return their objects.

2) Expose an `config: Dict[str, Any]` field (or use `**kwargs) on `StrawberryField` to be able to pass extra configurations to the field which the field extension can later read from. The big disadvantage here is that typing would not work very well.

Personally I like 1) a lot, but don't know what do you think about it.

cc @erikwrede

kaos reviewed (commented):
I still have an issue with using the simple
```python
@strawberry.type
class Query:
    node: relay.Node
    types: relay.Connection[TargetType]
```

which gives me a schema of
```gql
type Query {
  node: Node!
  types: TargetTypeConnection!
}
```

While using this relay connection works perfectly fine!:
```python
    @relay.connection
    def relay_tgt_types(self, info: Info, query: Optional[TargetTypesQuery] = None) -> Iterator[TargetType]:
        request_state = GraphQLContext.request_state_from_info(info)
        return TargetTypesQuery.filter(
            query,
            (
                TargetType.from_help(info)
                for info in request_state.all_help_info.name_to_target_type_info.values()
            ),
        )
```

I'll see if I can debug this further to isolate the root cause for this..

bellini666 commented on docs/guides/relay.md:
```diff
@@ -0,0 +1,396 @@
+We want it to have a globally unique ID, a way to retrieve a paginated results
+list of it and a way to refetch if if necessary. For that, we need to inherit it
+from the `Node` interface and implement its abstract methods: `resolve_id`,
+and `resolve_nodes`.
```
Nice catch!

bellini666 commented on docs/guides/relay.md:
```diff
@@ -0,0 +1,396 @@
+
+@strawberry.type
+class Fruit(relay.Node):
+    code: relay.NodeID[int]
```
Sure, I think it is nice to explain how `NodeID` is translated here.

But regarding the issue, I could not reproduce it. The tests in this PR are all using the annotation without the `= strawberry.field(...)`. Can you maybe give me a minimal reproduction example to try here?

kaos replied:
I'll see if can reproduce this in a simpler setup.

bellini666 commented:
> I still have an issue with using the simple

When using just the annotation without a resolver, it will use the type's `resolve_nodes` to get the nodes, and `nodes_id` will be `None` because it is asking for everything

mayteio commented:
What a rollercoaster ride reading the [issue started over a year ago](https://github.com/strawberry-graphql/strawberry/issues/1573) to now! The more the dates got more recent, the more excited I got! I appreciate your commitment to seeing this through. Pumped to see this released 🚀

jkimbo commented:
Btw @bellini666 and I had a discussion on discord about this PR (https://discord.com/channels/689806334337482765/1093176956524167258) and I've put forward a slightly tweaked API which you can see here: https://gist.github.com/jkimbo/3118b4bde65e67540d6cd45b9677cab4

Feedback welcome

bellini666 commented:
The PR is updated with the following changes:

1) It is now using field extensions instead of a custom field implementation

2) Lots of changes to the API based on [this feedback](https://gist.github.com/jkimbo/3118b4bde65e67540d6cd45b9677cab4) from @jkimbo . Regarding those:

- The `Connection` type needs to be explicit defined for each connection field together with a resolver. The connection type defines the pagination that will be used and the resolver the nodes to be paginated. For example:

```python
@strawberry.type
class Query:
    # It can be defined like this
    fruits: relay.ListConnection[Fruit] = relay.connection(resolver=get_fruits)

    # Or with a resolver
    @relay.connection(relay.ListConnection[Fruit])
    def fruits_with_resolver(self) -> Iterable[Fruit]:
        ...
```

- As you might have noticed above, I used `relay.ListConnection` instead of `relay.Connection`. The `Connection` is now an abstract implementation that cannot be used directly, but instead their subclasses should implement the pagination logic. `ListConnection` is one implementation provided by this integration that paginates nodes using a limit/offset approach.

- @jkimbo suggested to only have a `resolve_node` in the `Node` type to be defined by subclasses. I still kept the `resolve_nodes` as well because the integration supports not only `node: relay.Node` but also `nodes: List[relay.Node]`. When using a list you can ask for a list of node ids to be resolved (they don't even need to be of the same type) and their corresponding nodes will be returned.

I changed the `nodes_id` to not be optional anymore, which was the case of the previous integration. All you need to do in the `Node` subclass is to define a `resolve_nodes` classmethod for it to work. You can also override `resolve_node` for a custom, more performatic implementation. By default it calls `resolve_nodes` passing a list with a single element (the node id) to it.

3) I rewrote a good amount of the docs to reflect the changes in the API. I hope I covered most of the use cases. As always, I'm not the best at writing docs, so feel free to suggest whatever changes you find necessary :)

Hopefully we can get this merged soon! :)

cc @patrick91 @jkimbo @erikwrede

patrick91 commented on docs/guides/pagination/overview.md:
```diff
@@ -117,6 +117,14 @@ need a reliable and consistent way to handle pagination.
+[relay spec](https://relay.dev/docs/guides/graphql-server-specification/).
+You can read more about it in the [relay](./input-types) page.
+
+<Note/>
```
```suggestion
</Note>
```

nrbnlulu commented:
Thanks @bellini666 !!!

kaos commented:
amazing!!

woojing commented:
Great job! Thanks you @bellini666 and whole reviewers :)

patrick91 commented:
@lostclus can you open an issue for this? 

we were discussing some ideas around it on discord, but I think it is worth tracking it here on GitHub too :)
